pyat/at/acceptance/boundary.py

```diff
@@ -0,0 +1,407 @@
+                    istracked[i] = False
+                except KeyError:
+                    istracked[i] = True
+
+            ptmp = (part[:, istracked].T + offset).T
+            survived[istracked] = get_survived(ptmp, newring, nturns, method)
+
+            for i in range(len(angles)):
+                rp = get_r(part[planesi, i])
+                steps[i][rp] = survived[i]
+                if not survived[i] and rsteps[i] > rtol:
+                    part[planesi, i] -= cs[i]*min(2.0*rsteps[i], rp)
+                    rsteps[i] *= 0.5
+                    survived[i] = True
+
+        for pi, c in zip(planesi, cs.T):
+            part[pi] -= c*rsteps
+
+        grid = []
+        mask = []
+
+        for a, stp in zip(angles, steps):
+            for k, v in stp.items():
+                x = numpy.around(k*numpy.cos(a), decimals=9)
+                y = numpy.around(k*numpy.sin(a), decimals=9)
+                if v:
+                    mask.append([x, y])
+                grid.append([x, y])
+        p = numpy.squeeze(part[planesi])
+        m = numpy.array(mask).T
+        g = numpy.array(grid).T
+        return p, m, g
+
+    if use_mp:
+        method = at.patpass
+    else:
+        method = at.lattice_pass
+    offset, newring = set_ring_orbit(ring, dp, refpts, offset)
+    config = grid_configuration(planes, npoints, amplitudes,
+                                GridMode.RECURSIVE, bounds=bounds)
+    rtol = min(numpy.atleast_1d(config.amplitudes/config.shape))
+    rstep = config.amplitudes
+    if len(numpy.atleast_1d(config.shape)) == 2:
+        angles = numpy.linspace(*config.bounds[1], config.shape[1])
+    else:
+        angles = numpy.linspace(*config.bounds[1], 2)
+    angles = numpy.atleast_1d(angles)
+
+    if verbose:
+        print('\nRunning recursive boundary search:')
+        if refpts is None:
+            print('Element {0}, refpts={1}'.format(ring[0].FamName, 0))
+        else:
+            print('Element {0}, refpts={1}'.format(ring[refpts].FamName,
+                                                   refpts))
+        print('The grid mode is {0}'.format(config.mode))
+        print('The planes are {0}'.format(config.planes))
+        print('Number of angles is {0} from {1} to {2} rad'.format(len(angles),
+              angles[0], angles[-1]))
+        print('The resolution of the search is {0}'.format(rtol))
+        print('The initial step size is {0}'.format(rstep))
+        print('The initial offset is {0} with dp={1}'.format(offset, dp))
+
+    t0 = time.time()
+    result = search_boundary(ring, config.planesi, angles, rtol, rstep,
+                             nturns, offset, method)
+    if verbose:
+        print('Calculation took {0}'.format(time.time()-t0))
+    return result
+
+
+def boundary_search(ring, planes, npoints, amplitudes, nturns=1024,
+                    refpts=None, dp=None, offset=None, bounds=None,
+                    grid_mode=GridMode.RADIAL, use_mp=False, verbose=True):
```

Comment:
I would rename the `refpts` argument, as it is not the usual refpts type (integer,  array of integers or array of booleans). Here it is an index in the lattice.